### PR TITLE
Exploration with Parameter Space Noise

### PR DESCRIPTION
**Pull request by joneswong**
## What do these changes do?
Enable both DQN and DDPG to explore the environment with [parameter space noise](https://blog.openai.com/better-exploration-with-parameter-noise/)

## Related issue number
#3817

### CODE REVIEW

**Comment by joneswong** [comment]
Parameter space noise for exploration has been proved particularly effective in many continuous control problems. For instance, we used it for NIPS18 Prosthetics RL Challenge and made a comparison with traditional OU noise:
cmpexplorations

**Comment by joneswong** [comment]
I validated my implementation by comparing OU noise and parameter space noise in exploring pendulum:
pendulum_apex_ddpg

**Comment by joneswong** [comment]
As for discrete action space (DQN), I validated its usability and it is comparable with \epsilon greedy in exploring CartPole.

File: python/ray/rllib/agents/ddpg/ddpg_policy_graph.py [hunk 1]
```diff
@@ -429,6 +444,29 @@ def postprocess_trajectory(self,
...
+            distance_in_action_space = np.sqrt(
+                np.mean(np.square(clean_actions - noisy_actions)))
+            self.pi_distance = distance_in_action_space
+            if distance_in_action_space < self.config["exploration_sigma"]:
+                self.parameter_noise_sigma_val *= 1.01
+            else:
+                self.parameter_noise_sigma_val /= 1.01
+            self.parameter_noise_sigma.load(
+                self.parameter_noise_sigma_val, session=self.sess)
+
```
Comment (ericl) [inline python/ray/rllib/agents/ddpg/ddpg_policy_graph.py]: Nice!

File: python/ray/rllib/agents/dqn/dqn.py [hunk 2]
```diff
@@ -160,6 +167,12 @@ def _init(self):
             if k not in self.config["optimizer"]:
                 self.config["optimizer"][k] = self.config[k]
 
+        if self.config.get("parameter_noise", False):
+            self.config["callbacks"]["on_episode_start"] = tune.function(
+                on_episode_start)
```
Comment (ericl) [inline python/ray/rllib/agents/dqn/dqn.py]: This will override the callbacks the user sets right? Can we make sure to call the original ones afterwards?

**Review by ericl — COMMENTED** [review]
Could you also add an entry to run_multi_node_tests.sh to verify that param noise does not crash when enabled?

File: python/ray/rllib/agents/dqn/dqn.py (hunk 2 above)
Comment (joneswong) [reply python/ray/rllib/agents/dqn/dqn.py]: yes we must call them. I found that the callback functions are not allowed to include sth like self.config["callbacks"]["on_episode_start"]. They will cause pickle problem. I validated my implementation with some simple user specified callbacks.

Comment (ericl) [reply python/ray/rllib/agents/dqn/dqn.py]: Sounds good. I removed the .func from the start one if that's ok. It wasn't needed for the end one?

**Review by ericl — APPROVED** [review]

File: python/ray/rllib/agents/dqn/dqn.py (hunk 2 above)
Comment (joneswong) [reply python/ray/rllib/agents/dqn/dqn.py]: yes, actually no need for `.func`.

**Comment by joneswong** [comment]
It is difficult for me to find informative pieces from this log. There are many Errors. Could you give me some clues?

**Comment by ericl** [comment]
Merged, thanks!

**Comment by robertnishihara** [comment]
@ericl @joneswong this PR broke the linting on the master.

**Comment by ericl** [comment]
Weird, the Travis log showed a failure but I didn't  see any error. (exit 1)

**Comment by robertnishihara** [comment]
The command `cd doc; sphinx-build -W -b html -d _build/doctrees source _build/html` failed because this PR introduces a dependency on `scipy`. When readthedocs builds the documentation it needs to do `import ray`, and so we either need to add scipy to this list https://github.com/ray-project/ray/blob/acbe0b4e5f6cff6ce8fe8496d2c225b808dba9b1/doc/source/conf.py#L21 or to https://github.com/ray-project/ray/blob/master/doc/requirements-doc.txt (dependencies that are installed in order to build the documentation).

**Comment by robertnishihara** [comment]
Fixed in #4109.

**Pull request merged**